[HTTP 이해 - 웹 서버 구현] 2단계 미션 제출합니다.

bigpel66 opened the pull request:
안녕하세요 2단계 미션의 주요 내용은
* 임의의 접속에 JSESSIONID을 부여하고 이를 쿠키에 담아 관리
* 템플릿 엔진을 이용하여 /user/list에서 가입된 사용자의 목록을 나열
* Session과 SessionManager를 만들어 로그인 시 JSESSIONID에 Session을 맵핑하여 관리
* 로그인 된 상태에서 로그인 페이지 이동 시 index.html 이동
이었습니다.

이와 관련하여 동작 영상을 아래와 같이 첨부하였습니다.

https://github.com/next-step/jwp-was-kakao/assets/47660338/4f369048-b427-4c5c-af4d-3b9bf5ff098f

비고는 아래와 같습니다.
1. 템플릿 엔진의 경우 step1에서 추후 이용할 예정이라고 코멘트 드렸던 Attribute를 이용하였습니다.
2. Cookie 헤더에 담긴 구조와 Set-Cookie에 차이가 있다는 점 (Cookie는 한 줄에 나열되어 `; `로 구분된 값이 각 엔트리이고, Set-Cookie는 엔트리 등록을 하고 싶은 만큼 작성하여 브라우저로 넘겨야한다는 것)을 알게 되었습니다.
3. Set-Cookie를 이용하기 위해서 Path는 Key-Value 쌍 끝에 위치해야한다는 점도 알게되었습니다.
4. Session을 단순히 로그인 여부 판단을 위해 이용할 수 있다고만 생각했는데, 다양한 용도로 쓸 수 있겠다는 점을 인사이트로 얻었습니다.

이번에도 리뷰 진심으로 감사드립니다.

kriery commented on src/main/java/webserver/HttpRequest.java:
```diff
@@ -0,0 +1,153 @@
+            List<String> requestHeader = readRequestHeader(br);
+            Map<String, String> header = getHeaderFromRequestHeader(requestHeader);
+
+            int contentLength = header.containsKey("Content-Length") ? Integer.parseInt(header.get("Content-Length")) : 0;
```
반복되는 String-literal은 상수로 빼놓으면 좋을것 같아요

kriery commented on src/main/java/webserver/HttpRequest.java:
```diff
@@ -0,0 +1,153 @@
+    private static Map<String, String> getHeaderFromRequestHeader(List<String> requestHeader) {
+        Map<String, String> header = new HashMap<>();
+        for(String line: requestHeader) {
+            String[] data = line.split(": ");
```
(생각해볼 문제) 대부분의 browser에서는 요청 header에서 key-value delimiter를 `: `로 이용하고 있긴 하지만,
`:` 이후에 ` `(Space)는 0개 이상 출현 가능하긴 해서,
`KEY:    VALUE`나 `KEY:VALUE` 형태는 의도치 않은 결과를 초래할 수 있습니다.

kriery commented on src/main/java/webserver/HttpRequest.java:
```diff
@@ -0,0 +1,153 @@
+    private static List<String> readRequestHeader(BufferedReader br) throws IOException {
+        List<String> header = new ArrayList<>();
+        String line = br.readLine();
+        while (line != null && !"".equals(line)) {
```
이미 null check가 되어 있는 상태라서,
```suggestion
        while (line != null && !line.isEmpty()) {
```
해도 무방할것 같습니다. (isEmpty가 "".equals 보다 싸게 먹혀요)

kriery reviewed (approved):
안녕하세요 웨일론, WAS 미션 리뷰를 맡게된 크리스입니다.
기능별로 commit이 잘게 잘 나눠져 있고, 각 기능 구현/수정때 마다 테스트 케이스도 같이 변경되서 내용 따라가는데 큰 도움이 되었습니다 👍 
아무것도 없던 Controller가 점점 HttpServlet을 닮아가는 모습이 흥미로웠네요.
이대로 머지할게요, 몇가지 피드백을 드렸으니 확인해보고 다음 스탭에도 반영하면 좋을것 같습니다.

Commit:
feat : Cookie 객체 생성
- Cookie 객체를 구성하고 JSESSION이 정의됨을 확인
- GlobalControllerAdvice와 인터페이스는 ControllerAdvice로 정의
- CustomException을 Http Status Code에 따라 정의

Commit:
fix : Cookie 객체가 비어 있을 때 Set-Cookie가 동작하지 않도록 설정

Commit:
fix : 로그인 실패 시 리다이렉트를 수행

Commit:
feat : 로그인 시 User List를 확인할 수 있도록 반영
- Entry와 Cookie에서 사용하는 상수를 재배치
- GlobalControllerAdvice에서의 JSESSIONID 주입
- Set-Cookie 작성 오류 수정
- 템플릿 적용 안 된 스크립트를 수정

Commit:
feat : Session과 SessionManager를 구현

Commit:
feat : 로그인 후처리 구현
- 로그인 되어 있는 상태로 로그인 페이지 접근 시 index.html로 이동
- RequestLine을 수정하여 Service 재사용할 수 있도록 변경
- 템플릿 엔진에서 Attribute를 이용할 수 있도록 변경
- User List가 깨지는 문제를 수정

Commit:
docs : TODO 체크

sah3122 commented on src/main/java/webserver/WebApplicationServer.java:
```diff
@@ -31,6 +34,15 @@ public static void main(String[] args) throws Exception {
+            if (SessionManager.findSession(res.sessionId()).isPresent()) {
+                res.setMovedTemporarily(DEFAULT_ENTRY);
+                return;
+            }
```
로그인이 되어 있는 경우 페이지 이동을 처리해주셨네요 👍

sah3122 commented on src/main/java/webserver/common/Cookie.java:
```diff
@@ -0,0 +1,55 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public final class Cookie {
```
쿠키 객체 정의 👍

sah3122 commented on src/main/java/webserver/common/SessionManager.java:
```diff
@@ -0,0 +1,25 @@
+        SESSIONS.remove(id);
+    }
+
+    private SessionManager() {
+        throw new AssertionError("SessionManager 객체는 유틸리티 객체입니다.");
+    }
```
생성자는 클래스 최상단에 선언하곤 합니다 👍 
유틸성 클래스로 정의하여 생성을 막아주셨네요 👍

bigpel66 replied:
피드백 수정 반영 시 생성자 최상단에 위치시키도록 하겠습니다.

sah3122 commented on src/main/java/webserver/controller/GetUserListController.java:
```diff
@@ -0,0 +1,42 @@
+        res.setMovedTemporarily(LOGIN_ENTRY);
+    }
+
+    private boolean isLoggedIn(HttpResponse res) {
```
HttpResponse 객체를 통해 로그인 여부를 확인하기보단, HttpRequest 에 담겨있는 Cookie 를 통해 로그인 여부를 검증하는것은 어떨까요 ?

bigpel66 replied:
아래 코멘트를 달았는데, 말씀해주신 부분에 공감하는 바입니다. HttpRequest로 옮겨서 수행해보록 하겠습니다.

sah3122 commented on src/main/java/webserver/controller/GlobalControllerAdvice.java:
```diff
@@ -0,0 +1,37 @@
+            if (req.requiresInitialCookie(req.pathValue())) {
+                res.putCookie(JAVA_SESSION_KEY, UUID.randomUUID().toString());
+            }
+            controller.doService(req, res);
```
GlobalControllerAdvice를 두어 Proxy 역할을 할수 있도록 정의해주셨네요 👍 
AOP 와 비슷하게 동작하는 구조를 구현해주신것 같아 인상깊습니다 😄

sah3122 commented on src/main/java/webserver/controller/PostUserLoginController.java:
```diff
@@ -0,0 +1,52 @@
+                || !req.body().contains(passwordField);
+    }
+
+    private void recordLoginOnSession(HttpResponse res, User user) {
```
마찬가지로 response 객체를 활용해주신 이유가 있을까요 ?

bigpel66 replied:
HttpResponse를 이용하여 Set-Cookie의 렌더가 필요할 때는 HttpRequest에 있는 Cookie 값을 이용하게 되다보니 Cookie 값을 활용할 때 대체로 HttpResponse를 이용하는 형태로 구조가 잡힌 거 같습니다.

형태 상 HttpResponse를 이용하여 요청을 처리하는게 조금 어색하긴 한 거 같네요.

Cookie를 HttpRequest에 담아서 사용하도록 하고, 렌더 직전에 HttpResponse에 넘겨주는 식으로 구성해보는게 낫겠다 싶긴하네요.

sah3122 reviewed (changes requested):
2단계 미션 구현 잘해주셨습니다 👍 
몇가지 의문점이 있어 코멘트 남겨두었는데 확인 부탁드려요 🙏

Commit:
refactor : 2차 피드백을 반영
- Session 객체를 구성할 때 Request의 sessionId를 찾도록 변경
- SessionManager가 Session을 찾을 때 Response 대신 Request를 활용
- HttpResponse를 생성할 때 HttpRequest를 활용하지 않아 제거
- HttpResponse에 불필요한 주석과 컨벤션이 적용되지 않은 변수 이름 변경
- User List를 받아오는 서비스에서 로그인 여부를 판별하는 로직에서 Response가 아닌 Request를 이용

sah3122 commented on src/main/java/webserver/controller/GetUserListController.java:
```diff
@@ -18,7 +18,7 @@
     @Override
     public void doService(HttpRequest req, HttpResponse res) throws IOException {
-        if (isLoggedIn(res)) {
+        if (isLoggedIn(req)) {
```
HttpRequest를 활용하여 로그인 유무에 대한 검사구현 잘해주셨습니다 😄

sah3122 reviewed (approved):
피드백 반영 잘해주셨습니다 👍 
2단계 미션은 여기서 마무리 하도록 할게요 😄 
남아있는 미션들은 선택미션이니 진행을 원하신다면 리뷰요청 부탁드리겠습니다 !
